src/primitives.rs

```diff
@@ -136,3 +157,15 @@ impl FromByte for String {
 pub fn decode_buffer<F: FromByte>(buf: &mut Buf) -> Result<F, DecodeError> {
     FromByte::decode(buf)
 }
+
+pub trait FromByteWithVersion: Sized {
+    // ToDo: how to convert i16 to enum type ApiVersion?
```

Comment:
Got it. Then maybe I'll create a new function/method to convert `i16` to enum `ApiVersion` type.

Reply:
After a second thought, I don't think we need the enum `ApiVersion` here.

I found that if I convert `i16` to enum `ApiVersion` before calling the `decode_with_version()`, then I need to convert it back to `i16` inside `decode_with_version()` again to compare two version numbers as `i16`.

The Kafka messages **don't change that frequent** when the version number is upgraded. For example, for the `BodyMetadataRequest`, we have 

```
        if api_version <= 3 {
         ...
        } else {
            ...
        }
```

If we use enum `ApiVersion` instead, then we can either convert it back to `i16` to compare the version number with the threshold (version 3 in this case), or we can use `match` as the following:

```
        match api_version {
         Version0 | Version1 | Version2 => {
             ....
         },
         Version3 | Version4 | ... => {
            ...
        }
```

I don't think the enum `ApiVersion` is that helpful in this case. To avoid an invalidate version number (higher than the current latest version number, or is smaller than 0), we can check that when parsing the header rather than parsing the body.

Any idea?